community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java

```diff
@@ -674,21 +694,35 @@ State error( SessionStateMachine ctx, Neo4jError err )
         String connectionDescriptor();
         void reportError( Neo4jError err );
         void reportError( String message, Throwable cause );
-        KernelTransaction beginTransaction( KernelTransaction.Type type, AccessMode mode );
+        KernelTransaction beginTransaction( KernelTransaction.Type type, AccessMode mode,
+                                            TransactionIdTracker transactionIdTracker )
+                throws TransactionFailureException;
         void bindTransactionToCurrentThread( KernelTransaction tx );
         void unbindTransactionFromCurrentThread();
         RecordStream run( SessionStateMachine ctx, String statement, Map<String, Object> params )
                 throws KernelException;
         AuthenticationResult authenticate( Map<String, Object> authToken ) throws AuthenticationException;
         void udcRegisterClient( String clientName );
         Statement currentStatement();
+        TransactionIdTracker versionTracking( long startingVersion );
+
+        interface TransactionIdTracker
+        {
+            void assertUpToDate() throws TransactionFailureException;
```

Comment:
Looks strange that this method throws `TransactionFailureException`. Why assertion should know about transactions?
Maybe make this method return boolean and then use it to throw `TransactionFailureException` in `StandardStateMachineSPI#beginTransaction()`.

Also method name does not describe what it is doing. It should be smth like `waitUntilKnownTxOnParWithLastClosedTx`.